src/OrbitBase/include/OrbitBase/Executor.h

```diff
@@ -181,7 +181,14 @@ class Executor : public std::enable_shared_from_this<Executor> {
       // might do things that need synchronization and can't happen in a different context (like a
       // thread pool), i.e. when the continuation owns a ScopedStatus.
       if (argument.has_error()) {
+        // Here in SetResult other continuations might be called to propogate the result,
+        // and some of these continuations might result in calling of ScheduleAfter() which locks
+        // the same executor mutex. To avoid possible deadlocks, execute SetResult() here
+        // in the non-locked context.
+        executor_handle.data_->mutex_.ReaderUnlock();
         promise.SetResult(outcome::failure(argument.error()));
+        executor_handle.data_->mutex_.ReaderLock();
```

Comment:
When re-acquiring the lock we also have to check again whether the executor is still alive (similar to line 176) because the executor could have gone out of scope in the mean-time.

But I think the better approach would be to narrow the scope of the lock to the actual `ScheduleImpl` calls. As far as I can see this shouldn't have any detrimental side effects.

So here it would become:

```cpp
promise.SetResult(outcome::failure(argument.error()));

absl::ReaderMutexLock lock{&executor_handle.data_->mutex_};
if (executor_handle.data_->executor_ == nullptr) return;

executor_handle.data_->executor_->ScheduleImpl(CreateAction(
  [this, function_reference]() { waiting_continuations_.erase(function_reference); }));
return;
```

and down after the `success_function_wrapper` we would need the same.

WDYT?

Reply:
Yep, this was my idea. There is no way around having two locks I guess.

Reply:
Great, I'll fix it (most likely, today in the evening).

Reply:
Fixed.